dlmalloc/src/malloc.c

```diff
@@ -4139,13 +4147,23 @@ static void* sys_alloc(mstate m, size_t nb) {
             ssize < nb + SYS_ALLOC_PADDING) {
           size_t esize = granularity_align(nb + SYS_ALLOC_PADDING - ssize);
           if (esize < HALF_MAX_SIZE_T) {
+#ifdef __wasilibc_unmodified_upstream
             char* end = (char*)CALL_MORECORE(esize);
             if (end != CMFAIL)
               ssize += esize;
```

Comment:
it seems like an upstream bug to assume contiguity.
does dlmalloc still have a maintained upstream?

Reply:
dmalloc upstream has not been updated since 2012.

The comments at the top of dlmalloc say that it should handle the case of non-contiguous memory, though it will opportunistically detect contiguous memory and make use of it. If it has a bug, it'd be good to have a comment here mentioning that we're fixing an upstream bug.

Reply:
I don't think this is a bug, as it works correctly if dlmalloc is the only part of the program that uses sbrk (or MORECORE). That's just not necessarily the case in wasm where we want to allow other functions to grow the memory.

Reply:
> That's just not necessarily the case in wasm where we want to allow other functions to grow the memory.

Can you remind me where is this requirement coming from?  Would it not simplify thing considerably if were to declare it undefined behavior the call sbkr() or memory.grow on a memory that was owned/used/controlled by malloc?   i.e. can we not consider malloc as owning the entire region of memory from __heap_base to the end of the memory?

Reply:
I agree with giving malloc all the memory between `__heap_base` and `__heap_end` (which is not the end of the entire memory!), but I don't think we should prohibit other functions from growing the memory.

The use-case for me is to allow programs compiled with wasi-libc to have their own memory allocators in addition to malloc. When an application has very specific memory allocation patterns, such as all allocations having the same size, it is often more efficient to write a specialized allocator for a specific part of the program. On a regular linux target, you would then use mmap to allocate entire pages, so you don't need to mess with sbrk. However, in WebAssembly, sbrk/memory.grow is the only way to allocate new memory, so custom allocators need to able to use it without breaking malloc.

In any case, all memory pages allocated by malloc using memory.grow will continue to be used exclusively by malloc.

Reply:
> Can you remind me where is this requirement coming from?

tricky things like:

https://github.com/yamt/toywasm/blob/7f3610b2927e17f960233161596008298a0af6a2/examples/wasi_hook_c/hook-asm.S#L46-L55

https://github.com/bytecodealliance/wasm-tools/blob/0263c2f2e0a33fe97569535aa2784b7989cb1dd0/crates/wit-component/src/gc.rs#L99-L102

that is, someone wants to allocate from a linear memory of a wasi module, which doesn't export malloc/sbrk etc.

> Would it not simplify thing considerably if were to declare it undefined behavior the call sbkr() or memory.grow on a memory that was owned/used/controlled by malloc? i.e. can we not consider malloc as owning the entire region of memory from __heap_base to the end of the memory?

yea, it's simpler not to support these tricky use cases.
i'm personally fine with it.

Reply:
I do wonder if its worth it (or possible) to support that kind of thing.  

In that first example wouldn't make more sense if `mypage` was just statically allocated in the assembly file?    e.g:

```
.section .data.mypage,"",@ 
mypage:                                                                                                                                             
    .size mypage, 45536
```

Then the linker could place it and it would no runtime cost to allocate.

I'm not sure what the second example is used for, but I wonder if it couldn't depend on the `malloc` symbol existing? Or at least weakly depend on it so its used when it exists?

Reply:
> I do wonder if its worth it (or possible) to support that kind of thing.
> 
> In that first example wouldn't make more sense if `mypage` was just statically allocated in the assembly file? e.g:
> 
> ```
> .section .data.mypage,"",@ 
> mypage:                                                                                                                                             
>     .size mypage, 65536
> ```

no. because the linear memory is shared with another module.

> 
> Then the linker could place it and it would no runtime cost to allocate.
> 
> I'm not sure what the second example is used for, but I wonder if it couldn't depend on the `malloc` symbol existing? Or at least weakly depend on it so its used when it exists?

it's basically same as the first example.

i don't think it's possible unless malloc is exported by the module.

Reply:
a. do not support it (memory.grow outside of this malloc) at all
b. support it only if it's done at the very beginning of the module lifetime (eg. around wasm start function)
c. support it only if it's done before the first call of malloc
d. support it only if it's serialized with normal malloc/free calls
e. support it anytime, even from another thread

@MoritzS 
can't you make your use case backed by malloc as @sbc100 suggested? why not?
if you can't, which of the above do you need for your use case?

@alexcrichton @sunfishcode 
which does the preview1 adapter need?

Reply:
> > I do wonder if its worth it (or possible) to support that kind of thing.
> > In that first example wouldn't make more sense if `mypage` was just statically allocated in the assembly file? e.g:
> > ```
> > .section .data.mypage,"",@ 
> > mypage:                                                                                                                                             
> >     .size mypage, 65536
> > ```
> 
> no. because the linear memory is shared with another module.

If you have two modules sharing the same memory, isn't its reasonable to assume that one of them will export a malloc function and the other one will use it.   The need to coordinate somehow, and having them both call memory.grow and while assume that other can just deal with it seems.. fragile.   How do you know the allocator in other other module is ok with having a non-contiguous heap?  Isn't that fairly big assumption?  Why not just mandate that the other module export an allocator for you to use?

Reply:
> > > I do wonder if its worth it (or possible) to support that kind of thing.
> > > In that first example wouldn't make more sense if `mypage` was just statically allocated in the assembly file? e.g:
> > > ```
> > > .section .data.mypage,"",@ 
> > > mypage:                                                                                                                                             
> > >     .size mypage, 65536
> > > ```
> > 
> > 
> > no. because the linear memory is shared with another module.
> 
> If you have two modules sharing the same memory, isn't its reasonable to assume that one of them will export a malloc function and the other one will use it. The need to coordinate somehow, and having them both call memory.grow and while assume that other can just deal with it seems.. fragile. How do you know the allocator in other other module is ok with having a non-contiguous heap? Isn't that fairly big assumption? Why not just mandate that the other module export an allocator for you to use?

it isn't possible to assume proper export of malloc functions when you want to deal with the existing wasi modules.
i totally agree it's a fragile hack.

Reply:
>  which does the preview1 adapter need?

Sorry I haven't been keeping up with the rest of the context here, but the preview1 adapter won't ever be used with threads so it's ok if something odd goes wrong. Ideally wasi-libc would allow other modules to use `memory.grow` instructions themselves, however, and continue to work with that.